### PR TITLE
Gateway cached DataSource bug

### PR DESCRIPTION
**Pull request by trevor-scheer**
This PR addresses a bug in the gateway that prevents it from constructing new `DataSource`s when composition updates and the url of an underlying service changes. Currently the creation of a `DataSource` is cached based on its `serviceName`, but really the caching should depend on both the `serviceName` and `url` of the `ServiceDefinition`.

The bug in question was introduced in #3388.

Note:
This is broken out from its original PR: #3409
In order to separate two unrelated changes and cleanup some unnecessary git madness that I introduced.

### CODE REVIEW

File: packages/apollo-gateway/CHANGELOG.md [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 > The changes noted within this `vNEXT` section have not been released yet.  New PRs and commits which introduce changes should include an entry in this `vNEXT` section as part of their development.  When a release is being prepared, a new header will be (manually) created below and the the appropriate changes within that release will be moved into the new section.
 
-* _Nothing yet! Stay tuned!_
+* Gateway cached DataSource bug [#3412](https://github.com/apollographql/apollo-server/pull/3412)
```
Comment (zionts) [inline packages/apollo-gateway/CHANGELOG.md]: Could we include a more detailed message about what the bug was and what version it was introduced?

File: packages/apollo-gateway/src/index.ts [hunk 2]
```diff
@@ -262,9 +267,8 @@ export class ApolloGateway implements GraphQLService {
       return;
     }
 
-    if (!('serviceDefinitions' in result)) return;
-
     if (
+      !result.serviceDefinitions ||
```
Comment (zionts) [inline packages/apollo-gateway/src/index.ts]: Is this change intentional?

File: packages/apollo-gateway/src/index.ts [hunk 3]
```diff
@@ -379,8 +383,11 @@ export class ApolloGateway implements GraphQLService {
...
     // If the DataSource has already been created, early return
-    if (this.serviceMap[serviceDef.name])
-      return this.serviceMap[serviceDef.name];
+    if (
+      this.serviceMap[serviceDef.name] &&
+      serviceDef.url === this.serviceMap[serviceDef.name].url
+    )
+      return this.serviceMap[serviceDef.name].dataSource;
 
     if (!serviceDef.url && !isLocalConfig(this.config)) {
```
Comment (zionts) [inline packages/apollo-gateway/src/index.ts]: Should this really throw an error? Nothing that calls this actually handles an error effectively, so that means now that if you push up a bad version of your service with a `service:push`, the gateway will actually just crash rather than failing to update. Can we be more resilient to this instead?

Comment (zionts) [reply packages/apollo-gateway/src/index.ts]: Sorry, I realize this isn't new code, but I only just now noticed the strangeness of throwing an error within this function that isn't handled.

File: packages/apollo-gateway/src/loadServicesFromRemoteEndpoint.ts [hunk 4]
```diff
@@ -78,8 +78,5 @@ export async function getServiceDefinitionsFromRemoteEndpoint({
     serviceDefinitions.filter(Boolean),
   )) as ServiceDefinition[];
 
-  // XXX TS can't seem to infer that isNewSchema could be true
-  return (isNewSchema as true | false)
-    ? { serviceDefinitions, isNewSchema: true }
-    : { isNewSchema: false };
+  return { serviceDefinitions, isNewSchema }
```
Comment (zionts) [inline packages/apollo-gateway/src/loadServicesFromRemoteEndpoint.ts]: so much more pleasant.

File: packages/apollo-gateway/src/index.ts (hunk 2 above)
Comment (trevor-scheer) [reply packages/apollo-gateway/src/index.ts]: Yep! See changes from our previous PR. The previous lines were required for TS, but we don't actually want the early return there + the types are now permitting after the typing changes we make in this PR.
 
https://github.com/apollographql/apollo-server/pull/3409/files#diff-7b5e66976a637efc823826a985fa8e65R319

File: packages/apollo-gateway/CHANGELOG.md (hunk 1 above)
Comment (trevor-scheer) [reply packages/apollo-gateway/CHANGELOG.md]: Our changelog entries are typically pretty short with a link back to the PR.

Comment (zionts) [reply packages/apollo-gateway/CHANGELOG.md]: You know what best practices are better than I do, but I would imagine for major bugs that could affect your infrastructure _just not running_, we would have a more detailed changelog entry to encourage users who depend on the behavior to upgrade. 

From the perspective of someone who reads these changelogs / release notes often, I certainly would expect a more detailed reporting of major bugs and major new features. For instance, I wouldn't think `Gateway cached DataSource bug` would be something that causes my entire managed infrastructure to fail when my underlying services change URLs, but instead I would assume it was a bug around caching not being efficient enough.

File: packages/apollo-gateway/src/index.ts (hunk 3 above)
Comment (abernix) [reply packages/apollo-gateway/src/index.ts]: I haven't traced this precisely, but an unhandled exception in async functions or a rejected promise wouldn't typically result in a crash to the server, though it will certainly prevent the update from happening.  As far as I'm aware, the only current crash currently would be via the initial call to `createSchema` (which eventually arrives here) in the `ApolloGateway` constructor.   I believe the other way that this method is invoked is within async/Promise chains via `load`'s `setInterval`, and would fall into the category of non-terminating errors.

To be clear, a future version of Node.js will actually terminate the server under these conditions, but the brand new Node.js 12 LTS that is meant to arrive next week will still not take this approach

Either way, let's at least handle this now that we're aware of it though.

File: packages/apollo-gateway/src/index.ts [hunk 5]
```diff
@@ -495,17 +502,25 @@ export class ApolloGateway implements GraphQLService {
       }
     }
 
+    const serviceMap: ServiceMap = Object.entries(this.serviceMap).reduce(
+      (serviceDataSources, [serviceName, { dataSource }]) => {
+        serviceDataSources[serviceName] = dataSource;
+        return serviceDataSources;
+      },
+      {} as ServiceMap,
```
Comment (abernix) [inline packages/apollo-gateway/src/index.ts]: Let's use a `Object.create(null)` here so there's no prototype inheritance.  Otherwise, a `serviceName` of `hasOwnProperty` (or `constructor`, `__proto__`, etc.) might change, e.g., the otherwise expected dynamics of the object:

```suggestion
      Object.create(null) as ServiceMap,
```

File: packages/apollo-gateway/src/index.ts [hunk 6]
```diff
@@ -121,16 +125,17 @@ export type Experimental_DidUpdateCompositionCallback = (
   previousConfig?: Experimental_CompositionInfo,
 ) => void;
 
+/**
+ * Note: it's possible for a schema to be the same with serviceDefinitions that
+ * have changed. For example, during type migration.
```
[the remark this answers is not in the document]
Comment (trevor-scheer) [reply packages/apollo-gateway/src/index.ts]: Resolved in 57c9767 and 01ba2cb
